components/sup/src/main.rs

```diff
@@ -202,76 +202,85 @@ fn sub_term(m: &ArgMatches) -> Result<()> {
     }
     if let Some(addr_str) = m.value_of("LISTEN_HTTP") {
         cfg.http_listen = http_gateway::ListenAddr::from_str(addr_str)?;
     }
-    cfg.http_disable = m.is_present("HTTP_DISABLE");
     if let Some(addr_str) = m.value_of("LISTEN_CTL") {
-        cfg.ctl_listen =
-            SocketAddr::from_str(addr_str).unwrap_or_else(|_err| protocol::ctl::default_addr());
+        cfg.ctl_listen = SocketAddr::from_str(addr_str)?;
     }
-    cfg.organization = m.value_of("ORGANIZATION").map(|org| org.to_string());
-    cfg.gossip_permanent = m.is_present("PERMANENT_PEER");
-    // TODO fn: Clean this up--using a for loop doesn't feel good however an iterator was
-    // causing a lot of developer/compiler type confusion
-    let mut gossip_peers: Vec<SocketAddr> = Vec::new();
     if let Some(peers) = m.values_of("PEER") {
-        for peer in peers {
-            let peer_addr = if peer.find(':').is_some() {
-                peer.to_string()
-            } else {
-                format!("{}:{}", peer, GOSSIP_DEFAULT_PORT)
-            };
-            let addrs: Vec<SocketAddr> = match peer_addr.to_socket_addrs() {
-                Ok(addrs) => addrs.collect(),
-                Err(e) => {
-                    outputln!("Failed to resolve peer: {}", peer_addr);
-                    return Err(sup_error!(Error::NameLookup(e)));
-                }
-            };
-            let addr: SocketAddr = addrs[0];
-            gossip_peers.push(addr);
-        }
+        cfg.gossip_peers = get_peers(peers)?;
     }
-    cfg.gossip_peers = gossip_peers;
     if let Some(watch_peer_file) = m.value_of("PEER_WATCH_FILE") {
         cfg.watch_peer_file = Some(String::from(watch_peer_file));
     }
-    cfg.ring_key = match m.value_of("RING") {
-        Some(val) => Some(SymKey::get_latest_pair_for(
-            &val,
-            &default_cache_key_path(None),
-        )?),
+    if let Some(events) = m.value_of("EVENTS") {
+        cfg.eventsrv_group = ServiceGroup::from_str(events).ok().map(Into::into);
+    }
+    Ok(cfg)
+}
+
+// Various CLI Parsing Functions
+////////////////////////////////////////////////////////////////////////
+
+fn get_peers(peers: clap::Values) -> Result<Vec<SocketAddr>> {
+    // TODO fn: Clean this up--using a for loop doesn't feel good however an iterator was
+    // causing a lot of developer/compiler type confusion
+    let mut gossip_peers = Vec::new();
+    for peer in peers {
+        let peer_addr = if peer.find(':').is_some() {
+            peer.to_string()
+        } else {
+            format!("{}:{}", peer, GOSSIP_DEFAULT_PORT)
+        };
+        let addrs: Vec<SocketAddr> = match peer_addr.to_socket_addrs() {
+            Ok(addrs) => addrs.collect(),
+            Err(e) => {
+                outputln!("Failed to resolve peer: {}", peer_addr);
+                return Err(sup_error!(Error::NameLookup(e)));
+            }
+        };
+        let addr: SocketAddr = addrs[0];
```

Comment:
This is unsafe since according to the [`to_socket_addrs` docs](http://doc.rust-lang.org/1.29.2/std/net/trait.ToSocketAddrs.html#tymethod.to_socket_addrs):
> The returned iterator may not actually yield any values depending on the outcome of any resolution performed.

In general, it's best to avoid the [`index` operation](http://doc.rust-lang.org/1.29.2/std/ops/trait.Index.html#tymethod.index) (which `[…]` is sugar for) because it will panic if the index is invalid.

In this case, it's not too hard to do something a little safer:
```rust
            if let Some(addr) = addrs.get(0) {
                gossip_peers.push(*addr);
            }
```
Or you could avoid the unnecessary `collect` and creation of the `addrs` `Vec` with:
```rust
            match peer_addr.to_socket_addrs() {
                Ok(mut addrs) => if let Some(addr) = addrs.next() {
                    gossip_peers.push(addr);
                },
```

It also might be worth thinking a bit about in what cases `to_socket_addrs` *could* return either 0 or more than 1 results and what we want to do in those cases.